tachyon/crypto/hashes/sponge/poseidon/sparse_mds_matrix.h

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 Ethereum Foundation
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.EF and the LICENCE-APACHE.EF
+// file.
+
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON_SPARSE_MDS_MATRIX_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON_SPARSE_MDS_MATRIX_H_
+
+#include <string>
+#include <utility>
+
+#include "absl/strings/substitute.h"
+
+#include "tachyon/base/logging.h"
+#include "tachyon/base/strings/string_util.h"
+#include "tachyon/math/matrix/matrix_types.h"
+
+namespace tachyon::crypto {
+
+// +------+------+-----+------+   +-----+
+// | M₀,₀ | M₀,₁ | ... | M₀,ₜ | * | v₀  |
+// +------+------+-----+------+   +-----+
+// | M₁,₀ |                   |   | v₁  |
+// +------+                   +   +-----+
+// | ...  |         I         |   | ... |
+// +------+                   +   +-----+
+// | Mₜ,₀ |                   |   | vₜ  |
+// +------+------+-----+------+   +-----+
+//
+// |v[0]| = M₀,₀ * v₀ + M₀,₁ * v₁ + ... + M₀,ₜ * vₜ
+// |v[i]| = Mᵢ,₀ * v₀ + vᵢ, where 0 < i ≤ t
+template <typename F>
+class SparseMDSMatrix {
+ public:
+  SparseMDSMatrix() = default;
+  SparseMDSMatrix(const math::Vector<F>& row, const math::Vector<F>& col_hat)
+      : row_(row), col_hat_(col_hat) {
+    CHECK_EQ(row_.size(), col_hat_.size() + 1);
+  }
+  SparseMDSMatrix(math::Vector<F>&& row, math::Vector<F>&& col_hat)
+      : row_(std::move(row)), col_hat_(std::move(col_hat)) {
+    CHECK_EQ(row_.size(), col_hat_.size() + 1);
+  }
+
+  static SparseMDSMatrix FromMDSMatrix(const math::Matrix<F>& mds_matrix) {
+    CHECK_EQ(mds_matrix.rows(), mds_matrix.cols());
+    CHECK_GT(mds_matrix.rows(), 1);
+    CHECK_EQ(
+        mds_matrix.block(1, 1, mds_matrix.rows() - 1, mds_matrix.cols() - 1),
+        math::Matrix<F>::Identity(mds_matrix.rows() - 1,
+                                  mds_matrix.cols() - 1));
+    return {mds_matrix.row(0),
+            mds_matrix.block(1, 0, mds_matrix.rows() - 1, 1)};
+  }
+
+  const math::Vector<F>& row() const { return row_; }
+  const math::Vector<F>& col_hat() const { return col_hat_; }
+
+  void Apply(math::Vector<F>& v) const {
+    F v_0 = F::Zero();
+    for (Eigen::Index i = 0; i < v.size(); ++i) {
+      v_0 += row_[i] * v[i];
+    }
+
+    for (Eigen::Index i = 1; i < v.size(); ++i) {
+      v[i] += col_hat_[i - 1] * v[0];
+    }
```

Comment:
Isn't it faster to do it in parallel?

Reply:
I think usually mds matrix size is not that big, there's no merit to parallelize it.